src/libYARP_OS/include/yarp/os/Timer.h

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (C) 2017 iCub Facility
+ * Authors: Andrea Ruzzenenti
+ * CopyPolicy: Released under the terms of the LGPLv2.1 or later, see LGPL.TXT
+ */
+
+#include <yarp/os/RateThread.h>
+#include <yarp/os/Mutex.h>
+#include <functional>
+
+namespace yarp {
+    namespace os {
+        class Timer;
+        struct YarpTimerEvent;
+    }
+}
+
+struct yarp::os::YarpTimerEvent
+{   
+    /**
+     * @brief lastExpected when the last callback actually happened
+     */
+    double lastExpected;
+
+    /**
+     * @brief lastReal when the last callback actually happened
+     */
+    double lastReal;
+
+    /**
+     * @brief currentExpected this is when the current callback should have been called
+     */
+    double currentExpected;
+
+    /**
+     * @brief currentReal When the current callback is actually being called
+     */
+    double currentReal;
+
+    /**
+     * @brief lastDuration Contains the duration of the last callback
+     */
+    double lastDuration;
+
+    /**
+     * @brief runCount the count of calls
+     */
+    unsigned int runCount;
+};
+
+class YARP_OS_API yarp::os::Timer : public yarp::os::RateThread {
+    typedef std::function<void(const yarp::os::YarpTimerEvent&)> TimerCallback;
+
+    TimerCallback    m_callback{nullptr};
+    bool             m_oneshot{false};
+    double           m_startStamp{0.0},
+                     m_lastReal{0.0};
+    unsigned int     m_runCount{0};
+    yarp::os::Mutex* m_mutex;
+
+public:
+    /**
+     * @brief Timer constructor
+     * @param rate the call frequency, in milliseconds
+     * @param callback the pointer to the function to call
+     * @param mutex if not nullptr will be locked before calling callback and released just after
+     * @param oneshot if true there will be only one call
+     */
+    Timer(const double rate, TimerCallback callback, yarp::os::Mutex* mutex = nullptr, const bool oneshot = false);
+
+    /**
+     * @brief Timer constructor
+     * @param rate the call frequency, in milliseconds
+     * @param callback the pointer to the member method to call
+     * @param object the pointer to the object
+     * @param mutex if not nullptr will be locked before calling callback and released just after
+     * @param oneshot if true there will be only one call
+     */
+    template<class T>
+    Timer(const double rate,
```

Comment:
The `double` is usually used for "seconds" not "millisecond" in other YARP API, rate thread accepts `int` for "milliseconds", I'd like to remove any ambiguity in new API and use `std::chrono::milliseconds`, what do you think?